lambdajs/src/SharedInterpreter.hs

```diff
@@ -0,0 +1,183 @@
+{-# LANGUAGE Arrows                     #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+module SharedInterpreter where
+
+import           Prelude             hiding (break, lookup, map, read, fail)
+import qualified Prelude
+
+import           Control.Arrow
+import           Control.Arrow.Alloc
+import           Control.Arrow.Conditional as Cond
+import           Control.Arrow.Environment as Env
+import           Control.Arrow.Except as Exc
+import           Control.Arrow.Fail
+import           Control.Arrow.Fix
+import           Control.Arrow.Store as Store
+import           Control.Arrow.Utils
+
+import qualified Data.Label as Lab
+import           Data.String
+
+import           Text.Printf
+
+import           Syntax
+
+-- TODO
+-- This typeclass can be split up into multiple to partially reuse implementations across interpreters
+-- But this might not save effort if the implementation is completely different.
+
+class Arrow c => JSOps v env addr c | c -> v, c -> env, c -> addr where
+    -- simple values
+    numVal :: c Double v
+    boolVal :: c Bool v
+    stringVal :: c String v
+    undefVal :: c () v
+    nullVal :: c () v
+    evalOp :: c (Op, [v]) v
+
+    -- closures
+    closureVal :: c (env, [Ident], Expr) v
+    -- | applies a closure to an argument. The given continuation
+    -- describes how to evaluated the body of the closure.
+    applyClosure :: c ((env, [Ident], Expr),[v]) v -> c (v, [v]) v
+
+    -- objects
+    objectVal :: c [(Ident, v)] v
+    getField :: c (v, v) v
+    updateField :: c (v, v, v) v
+    deleteField :: c (v, v) v
+
+    -- store ops
+    ref :: c addr v
+    withRef :: c (e,(addr,v)) x -> c (e,v) x -> c (e,v) x 
+
+    -- control flow
+    label :: c Expr v -> c (Label, Expr) v
+    break :: c (Label, v) v
+
+withRef' :: (JSOps v env addr c, ArrowFail f c, IsString f, Show v) => c (e,(addr,v)) x -> c (e,v) x 
+withRef' f = withRef f (proc (_,refVal) -> fail -< fromString $ printf "Not a reference %s" (show refVal))
```

Comment:
If you want to use the new style with `withRef'`, you have to change the type of `withRef'` to `withRef' :: c (e,((addr,v),s)) x -> c (e,(v,s)) x`.

Then you can write
```Haskell
(| withRef' (\(a, _) -> read' -< a) |) refVal
```
and
```Haskell
(| withRef' (\(a, refVal)) -> do
                        write -< (a, rhsVal)
                        returnA -< refVal)
     |) refVal
```